contrib/dockerd-rootless.sh

```diff
@@ -139,6 +149,12 @@ if [ -z "$_DOCKERD_ROOTLESS_CHILD" ]; then
 		"$0" "$@"
 else
 	[ "$_DOCKERD_ROOTLESS_CHILD" = 1 ]
+
+	# Container Device Interface (CDI) specs can be found under {/etc,/var/run}/cdi.
+	# More information at: https://github.com/cncf-tags/container-device-interface
+	[ -d "/etc/cdi" ] && mount_directory /etc/cdi
```

Comment:
Question: is this run once at startup, or once per container? I would expect NEW specs added to these folders to be accessible to containers without requiring a restart of the rootless daemon.

Reply:
This is executed once when the Dockerd daemon is started. However, the Dockerd daemon will be watching the default CDI locations, so that if new specs are added either to `/etc/cdi` or `/var/run/cdi`, they will be available to the daemon without the need to restart it.

Just for the sake of an example, I just checked it by copying my `/var/run/cdi/nvidia-container-toolkit.json` to `/etc/cdi` modifying the `kind` to `somevendor.com/tpu`. Without any daemon restart, I can do:

```
❯ DOCKER_HOST=unix:///run/user/1000/docker.sock docker run --rm --device=somevendor.com/tpu=all -it ubuntu:latest nvidia-smi -L
GPU 0: NVIDIA GeForce RTX 4090 (UUID: GPU-c475e08b-0cc5-f5aa-4326-99699429b449)
GPU 1: NVIDIA GeForce RTX 2080 SUPER (UUID: GPU-5cca1a6f-7cee-b649-40f0-2d3ecb0aa207)
```

Reply:
Sorry, I should have been clearer. What is the expectation if the folders *do not exist* when the rootless daemon is started (we explicitly have a `-d` conditional guard)? In the case of the rootfull daemon, I believe the underlying CDI code will detect the creation of the configured directories and refresh the internal CDI cache based on the contents added.

We should at least document that for the CDI injection to work as expected, the `/etc/cdi` and / or `/var/run/cdi` folders needs to exist when the rootless daemon is started.

Reply:
Another question: The CDI spec directories are configurable. Should we adjust the behaviour of the rootless daemon based on that config or is it ok to always assume the defaults?

Reply:
Good points @elezar. Thanks!

> What is the expectation if the folders do not exist when the rootless daemon is started (we explicitly have a -d conditional guard)? 

Yes, in this case, they won't be included and creating files under them will have no effect without a restart of the daemon.

> We should at least document that for the CDI injection to work as expected, the /etc/cdi and / or /var/run/cdi folders needs to exist when the rootless daemon is started.

I have updated the script with a explicit mention to this behavior as a comment. Please, let me know if you were thinking in other kind of documentation, more public. I'd be happy to include that too.

> Another question: The CDI spec directories are configurable. Should we adjust the behaviour of the rootless daemon based on that config or is it ok to always assume the defaults?

I think it's better to keep it simple. Using the config would mean including `jq` as another dependency and also making sure we are processing the right JSON. Adding something middle ground such a new envvar would give a wrong signal to the user (do I need to provide the envvar AND the cdi-spec-dirs in the config? what happens if they don't match?). For these reasons, I think it's better to keep it simple for now, until we have some feedback at least that signals otherwise.

Reply:
I think the comment is fine for now. I'm also happy to put off trying to create this folder or monitor for its existence for the time being. Let's rather get feedback on the use of CDI in this context before making things more public or more complex.